test/nodetool/rest_api_mock.py

```diff
@@ -0,0 +1,149 @@
+#
+# Copyright 2023-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+import http.server
+import json
+import requests
+import socketserver
+import sys
+
+
+class request:
+    def __init__(self, method, path, response=None):
+        self.method = method
+        self.path = path
+        self.response = response
+
+    def as_json(self):
+        res = {"method": self.method, "path": self.path}
+        if self.response is not None:
+            res["response"] = self.response
+        return res
+
+    def __eq__(self, o):
+        return self.method == o.method and self.path == o.path
+
+    def __str__(self):
+        return json.dumps(self.as_json())
+
+
+def _make_request(req_json):
+    return request(req_json["method"], req_json["path"], req_json.get("response"))
+
+
+class response_handler(http.server.BaseHTTPRequestHandler):
+    EXPECTED_REQUESTS_PATH = "/__expected_requests__"
+    expected_requests = []
+
+    def set_headers(self, content_length):
+        self.protocol_version = "HTTP/1.1"
+        self.send_response(200)
+        self.send_header("Content-Length", str(content_length))
+        self.send_header("Content-Type", "application/json")
+        self.end_headers()
+
+    def do_handle(self, method):
+        this_req = request(method, self.path)
+
+        if len(response_handler.expected_requests) == 0:
+            self.send_error(500, message="Unexpected request", explain=f"Expected no requests, got: {this_req}")
+            return
+
+        expected_req = response_handler.expected_requests[0]
+        if this_req != expected_req:
+            self.send_error(500, message="Unexpected request", explain=f"Expected {expected_req}, got: {this_req}")
+            return
+
+        if expected_req.response is None:
+            self.log_message(f"expected_request: {expected_req}, no response")
+            self.set_headers(0)
+        else:
+            self.log_message(f"expected_request: {expected_req}, response: {expected_req.response}")
+            payload = expected_req.response.encode()
+            self.set_headers(len(payload))
+            self.wfile.write(payload)
+
+        del response_handler.expected_requests[0]
+
+    def do_GET(self):
+        self.close_connection = False
```

Comment:
Why do this only for do_GET() and not for all kinds of requests?

Setting close_connection=False should be fine as long as you make sure to read the body of the request (for GET, it's usually 0 bytes, but doesn't have to). Thinking about it, if you had to set close_connection=False only here for GET but didn't have to do the same for for POST, maybe the library set it to true just in do_GET before you forgot to read the body in do_GET (but did get it to do_POST)?

Reply:
If I set this for all kind of requests, then the tests hung. Maybe reading the body of the request (even if 0) will solve this. If not, then I will have to have some logic to determine whether more requests are expected or not.

Reply:
`http.server` was thrown out altogether in v4.